### PR TITLE
Partitioned heat conduction with Schwarz-type domain decomposition

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
Adds a tutorial illustrating how Schwarz-type domain decomposition could be implemented with preCICE.

General question similar to #30: How can we publish these kind of academic and advanced tutorials in a proper way?

### CODE REVIEW

File: partitioned-heat-conduction-schwarz/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{% endnote %}
+
+## Running the simulation
+
+This tutorial is for FEniCS.
+
+For choosing whether you want to run the left or right participant, please provide the following commandline input:
+
+* `heat.py left` flag will run the left participant.
+* `heat.py right` flag will run the right participant.
```
Comment (uekerman) [inline partitioned-heat-conduction-schwarz/README.md:19]: Could you briefly mention the error computation script here and what to expect?

**Review by uekerman — COMMENTED** [review]
Could we add a picture showing the overlapping domain? I think this could be very helpful to quickly grasp the difference of this tutorial to others.

> How can we publish these kind of academic and advanced tutorials in a proper way?

This discussion will be part of the preECO project. Let's not preempt a decision here. For the moment, what you did looks good.

I did not run the case.

File: partitioned-heat-conduction-schwarz/README.md (hunk 1 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-schwarz/README.md]: I added a few lines. It's the same like for the Dirichlet-Neumann case: If we do everything right, we are able to approximate the analytical solution up to machine precision (or coupling tolerance).

File: partitioned-heat-conduction-overlap/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+---
+title: Partitioned heat conduction with overlapping Schwarz-type domain decomposition
+permalink: tutorials-partitioned-heat-conduction-overlap.html
+keywords: FEniCS, Nutils, Heat conduction
```
Comment (uekerman) [inline partitioned-heat-conduction-overlap/README.md:4]:
```suggestion
keywords: FEniCS, heat conduction, overlap
```

File: partitioned-heat-conduction-overlap/fenics/heat.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+# Heat equation with mixed boundary conditions. (Neumann problem, $u_C = \mathcal{N}\left(f_N\right)$)
+#
+# \begin{align*}
+# \frac{\partial u}{\partial t} &= \Delta u + f && \text{on the shifted unit square $\Omega = \left[1,2\right] \times \left[0,1\right]$}\\
+# \frac{\text{d}u}{\text{d}\vec{n}} &= f_N             && \text{on the coupling boundary $\Gamma_C$ at $x = 1$}\\
+# u &= u_D             && \text{on the remaining boundary $\Gamma = \partial \Omega \setminus \Gamma_C$}\\
+# u &= u_0             && \text{on $\Omega$ at t = 0}\\
+# u &= 1 + x^2 + \alpha y^2 + \beta t \\
+# f &= \beta - 2 - 2\alpha \\
+# \end{align*}
```
Comment (uekerman) [inline partitioned-heat-conduction-overlap/fenics/heat.py:32]: needs updating

File: partitioned-heat-conduction-overlap/precice-config.xml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  <log>
+    <sink
+      filter="%Severity% > debug and %Rank% = 0"
+      format="---[precice] %ColorizedSeverity% %Message%"
+      enabled="true" />
+  </log>
+
+  <solver-interface dimensions="2">
+    <data:scalar name="TemperatureLeft" />
+    <data:scalar name="TemperatureRight" />
```
Comment (uekerman) [inline partitioned-heat-conduction-overlap/precice-config.xml:12]:
```suggestion
    <data:scalar name="Temperature-Left" />
    <data:scalar name="Temperature-Right" />
```
would be slightly more consistent

File: partitioned-heat-conduction-overlap/images/tutorials-partitioned-heat-conduction-overlap-setup.png
Comment (uekerman) [inline partitioned-heat-conduction-overlap/images/tutorials-partitioned-heat-conduction-overlap-setup.png:1]: A bit odd to call an overlap region "Gamma", but OK.

**Review by uekerman — APPROVED** [review]
Name and README look good enough for me. I did not run the case.

File: partitioned-heat-conduction-overlap/fenics/heat.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+# set t_1 = t_0 + dt, this gives u_D^1
+# call dt(0) to evaluate FEniCS Constant. Todo: is there a better way?
+u_D.t = t + dt(0)
+f.t = t + dt(0)
+
+while precice.is_coupling_ongoing():
+
+    # write checkpoint
+    if precice.is_action_required(precice.action_write_iteration_checkpoint()):
```
Comment (MakisH) [inline partitioned-heat-conduction-overlap/fenics/heat.py:209]: Is this using the FEniCS adapter for preCICE v3?

File: partitioned-heat-conduction-overlap/fenics/precice-adapter-config-L.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+  "participant_name": "Left",
+  "config_file_name": "../precice-config.xml",
+  "interface": {
+      "coupling_mesh_name": "Left-Mesh",
+      "write_data_name": "TemperatureLeft",
+      "read_data_name": "TemperatureRight"
```
Comment (MakisH) [inline partitioned-heat-conduction-overlap/fenics/precice-adapter-config-L.json:7]: This is classical Dirichlet-Neumann, right? Strange to call both exchanged fields `Temperature*`. A gradient is needed somewhere.

File: partitioned-heat-conduction-overlap/images/tutorials-partitioned-heat-conduction-overlap-setup.tex
Comment (MakisH) [inline partitioned-heat-conduction-overlap/images/tutorials-partitioned-heat-conduction-overlap-setup.tex:1]: Nice that you added the source file of the figure here!

File: partitioned-heat-conduction-overlap/precice-config.xml [hunk 7]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<precice-configuration>
+  <log>
+    <sink
+      filter="%Severity% > debug and %Rank% = 0"
+      format="---[precice] %ColorizedSeverity% %Message%"
+      enabled="true" />
+  </log>
+
+  <solver-interface dimensions="2">
```
Comment (MakisH) [inline partitioned-heat-conduction-overlap/precice-config.xml:10]: Needs to be ported to preCICE v3

File: partitioned-heat-conduction-overlap/fenics/heat.py
Comment (MakisH) [inline partitioned-heat-conduction-overlap/fenics/heat.py:1]: We have already discussed this in other occasions, but this is another example of a tutorial that has a directory structure based on the solver and not on the participant.

An easy split here would be:

- `fenics-solver`:
   - `heat.py`
   - `errorcomputation.py`
- `left-fenics`:
   - `precice-adapter-config-left.json`
   - `run.sh` (without options)
   - `clean.sh`   
- `right-fenics`:
   - `precice-adapter-config-right.json`
   - `run.sh` (without options)
   - `clean.sh`

**Review by MakisH — CHANGES_REQUESTED** [review]
I have not tried to run this tutorial, but it anyway needs porting to preCICE v3.

I would also prefer that it already applies the tutorial structure that we have essentially discussed and converged on, which is documented in https://precice.org/community-contribute-to-precice.html

Other than that, it looks clean and clear. I particularly like that you added the TikZ sources for the figure.

File: partitioned-heat-conduction-overlap/precice-config.xml (hunk 4 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/precice-config.xml:12]: Done. 91d552d9af2170a5175309e465eaf6f7ced9ea2b

File: partitioned-heat-conduction-overlap/images/tutorials-partitioned-heat-conduction-overlap-setup.png
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/images/tutorials-partitioned-heat-conduction-overlap-setup.png:1]: The overlap region is the shaded region and called $\Omega_\mathcal{L} \cap \Omega_\mathcal{R}$. $\Gamma_\mathcal{L}$ and $\Gamma_\mathcal{R}$ are only the boundaries where we set the Dirichlet BCs. But I moved the labels a bit to make things clearer.

File: partitioned-heat-conduction-overlap/fenics/heat.py (hunk 5 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/fenics/heat.py:209]: No. I think this has just been lying around too long.

File: partitioned-heat-conduction-overlap/fenics/precice-adapter-config-L.json (hunk 6 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/fenics/precice-adapter-config-L.json]: No. It is Schwarz. Here, you have an overlap region and exchange Dirichlet BCs. See https://en.wikipedia.org/wiki/Schwarz_alternating_method

File: partitioned-heat-conduction-overlap/precice-config.xml (hunk 7 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/precice-config.xml]: Done :+1:

File: partitioned-heat-conduction-overlap/fenics/heat.py
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/fenics/heat.py:1]: Done :+1:

File: partitioned-heat-conduction-overlap/fenics/heat.py (hunk 5 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/fenics/heat.py]: Done with update.

File: partitioned-heat-conduction-overlap/fenics/heat.py (hunk 3 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/fenics/heat.py]: Done.

File: partitioned-heat-conduction-overlap/fenics/precice-adapter-config-L.json (hunk 6 above)
Comment (MakisH) [reply partitioned-heat-conduction-overlap/fenics/precice-adapter-config-L.json]: Hmmm.... I was probably (mentally) absent during the respective NumPro lecture... :see_no_evil:

File: partitioned-heat-conduction-overlap/README.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+## Running the simulation
+
+This tutorial is for FEniCS.
+
+For choosing whether you want to run the left or right participant, please provide the following commandline input:
+
+* `python3 heat.py left` flag will run the left participant.
+* `python3 heat.py right` flag will run the right participant.
+
+Like for the case `partitioned-heat-conduction` (using Dirichlet-Neumann coupling), we can also expect for the overlapping domain decomposition applied here to recover the analytical solution. `errorcomputation.py` checks this explicitly, by comparing the numerical to the analytical solution and raising an error, if the approximation error is not within a given tolerance.
```
Comment (MakisH) [inline partitioned-heat-conduction-overlap/README.md:27]:
```suggestion
Like for the case `partitioned-heat-conduction` (using Dirichlet-Neumann coupling), we can also expect for the overlapping domain decomposition applied here to recover the analytical solution. The script `solver-fenics/errorcomputation.py` checks this explicitly, by comparing the numerical to the analytical solution and raising an error, if the approximation error is not within a given tolerance.
```

File: partitioned-heat-conduction-overlap/solver-fenics/errorcomputation.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+def compute_errors(u_approx, u_ref, v, total_error_tol=10 ** -4):
+    # compute pointwise L2 error
+    error_normalized = (u_ref - u_approx) / u_ref
+    # project onto function space
+    error_pointwise = project(abs(error_normalized), v)
+    # determine L2 norm to estimate total error
+    error_total = sqrt(assemble(inner(error_pointwise, error_pointwise) * dx))
+    error_pointwise.rename("error", " ")
+
+    assert (error_total < total_error_tol)
```
Comment (MakisH) [inline partitioned-heat-conduction-overlap/solver-fenics/errorcomputation.py:13]: Would it make sense to still print a message on success?
I ran this script, semi-following the documentation, but I was not sure what the absence of output meant.

**Review by MakisH — APPROVED** [review]
The new structure is nice! :+1:
I ran the code, no VTK files are produced (as expected, I supposed).
I then executed the `solver-fenics/errorcomputation.py` and got no output, which confused me but means that the assertion worked.

File: partitioned-heat-conduction-overlap/README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+{% note %}
+Get the [case files of this tutorial](https://github.com/precice/tutorials/tree/master/partitioned-heat-conduction-overlap). Read how in the [tutorials introduction](https://www.precice.org/tutorials.html).
+{% endnote %}
+
+## Setup
+
+We solve a partitioned heat equation, but apply an overlapping Schwarz-type domain decomposition method in this tutorial.
+
+![Case setup of partitioned-heat-conduction case with overlapping Schwarz-type domain decomposition](images/tutorials-partitioned-heat-conduction-overlap-setup.png)
+
```
Comment (MakisH) [inline partitioned-heat-conduction-overlap/README.md:17]: The clarification you have in the oscillator-overlap tutorial would also be helpful here:

```suggestion

Note that this case applies an overlapping Schwarz-type coupling method and not (like most other tutorials in this repository) a Dirichlet-Neumann coupling. This results in a symmetric setup of the solvers.

```

You can even extend this explaining how you name the participants and that they both exchange temperature.

File: partitioned-heat-conduction-overlap/solver-fenics/errorcomputation.py (hunk 9 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction-overlap/solver-fenics/errorcomputation.py:13]: Not sure. I think it is ok the way it is (we also have it this way for the other heat conduction flavors).

**Comment by BenjaminRodenberg** [comment]
> I ran the code, no VTK files are produced (as expected, I supposed).

I don't completely understand this. I get VTK output...?

Screenshot from 2024-02-05 23-27-53

File: partitioned-heat-conduction-overlap/solver-fenics/errorcomputation.py (hunk 9 above)
Comment (MakisH) [reply partitioned-heat-conduction-overlap/solver-fenics/errorcomputation.py:13]: I am a user.
I got confused.
The _"we also have it this way for the other heat conduction flavors"_ does not mean that I wouldn't get confused in those as well... :see_no_evil:

**Comment by MakisH** [comment]
> > I ran the code, no VTK files are produced (as expected, I supposed).
> 
> I don't completely understand this. I get VTK output...?

I guess I overlooked the `output/` folder, used to having the Nutils output in the same directory. But this is not our Nutils tutorials.

**Pull request merged**